File: src/com/train/algorithm/tree/implementInC++/DistributeCoinsinBinaryTree.cpp
//
// Created by 罗斯 on 14/11/2022.
//
#include <algorithm>

using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int distributeCoins(TreeNode* root) {
        int res = 0;
        balance(root, res);
        return res;
    }
private:
    int balance(TreeNode* root, int& res){
        if (!root) return 0;
        int l = balance(root->left, res);
        int r = balance(root->right, res);
        res += abs(l) + abs(r);
        return l + r + root->val - 1;
    }
};

File: src/com/train/algorithm/tree/implementInPython/DistributeCoinsInBinaryTree.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def distributeCoins(self, root):
        self.moves = 0
        def helper(node):
            if not node: return 0
            left = helper(node.left)
            right = helper(node.right)

            cur = left + right + node.val - 1
            need = abs(left) + abs(right)

            self.moves += need
            return cur

        helper(root)
        return self.moves